ambry-quota/src/main/java/com/github/ambry/quota/MySqlStorageUsageRefresher.java

```diff
@@ -0,0 +1,470 @@
+        containerStorageUsageMonthlyBase = backupFileManager.getBackupFileContent(currentMonth);
+      }
+      if (containerStorageUsageMonthlyBase != null) {
+        return;
+      }
+    } catch (IOException e) {
+      logger.error("Failed to get container monthly usage for " + currentMonth + " from backup", e);
+    }
+
+    try {
+      // If we are here, then loading monthly base from backup file failed. We have to fetch it from database.
+      containerStorageUsageMonthlyBase =
+          accountStatsMySqlStore.queryMonthlyAggregatedStats(clusterMapConfig.clusterMapClusterName);
+      // If the monthly base is indeed for this month, then try to persist it in the backup file.
+      // There is a chance that the database has a snapshot from last month since the aggregation task is executed
+      // every few minutes(maybe hours). Before the first aggregation task of this month is executed, the database
+      // would have the snapshot of last month.
+      if (currentMonth.equals(accountStatsMySqlStore.queryRecordedMonth(clusterMapConfig.clusterMapClusterName))) {
+        tryPersistMonthlyUsage();
+      }
+    } catch (Exception e) {
+      throw new IllegalStateException("Unable to fetch monthly storage usage from mysql", e);
+    }
+  }
+
+  /**
+   * Try to persist the monthly base usage in the backup file and ignore any exceptions.
+   */
+  private void tryPersistMonthlyUsage() {
+    if (backupFileManager != null) {
+      try {
+        backupFileManager.persistentBackupFile(currentMonth, containerStorageUsageMonthlyBase);
+      } catch (IOException e) {
+        // Error already been logged from backup file manager.
+      }
+    }
+  }
+
+  /**
+   * Fetch the container monthly storage usage from mysql and update the in memory cache. If somehow it fails, we will
+   * retry for several times before giving up.
+   */
+  private void fetchMonthlyStorageUsageAndMaybeRetry() {
+    boolean shouldRetry = false;
+    try {
+      String monthValue = accountStatsMySqlStore.queryRecordedMonth(clusterMapConfig.clusterMapClusterName);
+      if (monthValue.equals(currentMonth)) {
+        Map<String, Map<String, Long>> base =
+            accountStatsMySqlStore.queryMonthlyAggregatedStats(clusterMapConfig.clusterMapClusterName);
+        containerStorageUsageMonthlyBase = base;
+      } else {
+        shouldRetry = true;
+      }
+    } catch (Exception e) {
+      logger.error("Failed to refresh monthly storage usage", e);
+      shouldRetry = true;
+    }
+
+    if (shouldRetry) {
+      if (retries >= config.mysqlStoreRetryMax) {
+        logger.error("Failed to refresh monthly storage usage after " + retries + " retries, will skip for month: "
+            + currentMonth);
+        retries = 0;
+      }
+      retries++;
+      if (config.mysqlStoreRetryBackoffMs != 0) {
+        scheduler.schedule(this::fetchMonthlyStorageUsageAndMaybeRetry, config.mysqlStoreRetryBackoffMs,
+            TimeUnit.MILLISECONDS);
+      }
```

Comment:
Could you explain the logic here?  Looks like if `retries` exceeds max count, it resets `retries` , increments its value and  schedule another round of retry ?

Reply:
I have made a mistake. I should schedule for next fetch after resetting the retries back to 0 and return.

The logic goes like this, assume max retry count is 3. Then after starting to fetching the monthly base, we can try it for 3 times. If for all this 3 times, it failed, then we would just schedule for next fetch.